Avoid ambiguous block heights as much as possible in masternode code

domob1812 opened the pull request:
A lot of the masternode code is using block heights for things like computing the payment winners or even as keys into caches and other in-memory maps.  This is problematic, as a block height may refer to more than one actual block in the case of reorgs.

In this set of changes, we change many of those to be based on block hashes.  For payment winners in particular, what matters is the *hash of the block 101 blocks earlier* (`GetBlockHashForScoring`, previously `GetBlockHash`).  We use this seed / scoring hash as main piece of data for keying winner-related data.

For actually computing the winner hash, we now use most of the time an explicit `CBlockIndex*` as basis, rather than `chainActive` (i.e. explicitly specify the chain in which we look for the ancestor at the required height).  This avoids using `chainActive` as a global, and makes the code more robust against reorgs in general.

The only place where we still use only block heights is the wire format for `CMasternodePaymentWinner` messages (i.e. processing a `mnw` received from a peer, as well as the hash used to identify them in a node's inventory).  Changing this to the seed hash would change the wire protocol, and thus make the change much more complex to do.

Finally, this also includes a simplification and potential fix to how votes for winners are tracked and duplicate votes by a single masternode prevented.

domob1812 committed:
Simplify masternode GetBlockHash.

The GetBlockHash function in the masternode code is quite complex
for what it does (look up a block hash by height).  This is a first
simplification step, which essentially just removes dead parts of
the code and is thus trivial to review.

domob1812 committed:
Remove "mod" for CalculateScore.

CalculateScore was not using the "mod" in any way, and all callers
were passing "1".  Remove it to clarify the code, and also make the
other arguments mandatory to ensure there is no accidental confusion
about which arguments are passed.

domob1812 committed:
Rename masternode GetBlockHash function.

The masternode.cpp GetBlockHash function does something slightly
different than getting the block hash for a height (namely it gets
the hash for the height - 1, mostly).

To avoid any confusion and make it clear where this specific function
is used, rename it to GetMnBlockHash.

domob1812 committed:
Simplify GetMnBlockHash.

All that GetMnBlockHash really does is look up the block hash at
the given height, or actually one before.  This can be trivially done
through chainActive, and needs neither a cache (which is actually also
a risk due to reorgs) nor iterating through "blocks ago".

domob1812 committed:
Remove GetCurrentMasterNode.

Get rid of GetCurrentMasterNode.  The function was only being used in
one place, namely as fallback for filling the payee of a created block
when there are no votes to use instead.  And in this place, it was being
called with the wrong height (it should have been offset with -100 as in
any other place calculating scores).

Since we already have a correct function to get the masternode to pay to
with GetNextMasternodeInQueueForPayment, just use this, which also takes
care of the offset already.

domob1812 committed:
Pass explicit seed hash to CalculateScore.

Instead of calculating the score of a masternode based on a block
height, pass the underlying block hash (which is used to seed the
score randomness) directly.  This makes the code more explicit, and
will allow better separation of the translation from height to hash
(which is somewhat prone to reorg issues).

Also moves the -100 block height offset into GetBlockHashForScoring
(GetMnBlockHash renamed), so that this is only specified in a single
place.  The function itself is now passed the block height for which
we want the winner / score (not the offset height).

domob1812 committed:
Pass scoring hash to GetMasternodeRank.

Instead of passing the block height and computing the scoring hash from
it inside GetMasternodeRank, pass the scoring hash directly into the
function (and compute it outside of it at the callsites).

This also allows us to change the ranking cache to be keyed by scoring
hash instead of block height, which makes it immune against potential
issues due to reorgs.

domob1812 committed:
Store scoring hash with CMasternodePaymentWinner.

In each CMasternodePaymentWinner message, store the scoring hash
in addition to the block height.  This is what ultimately determines
the winner, and is thus the "proper key" for those votes (while the
height is prone to reorgs).

Since the wire protocol serialises (and hashes) these with height and
we do not want to change the protocol, we only compute the seed hash
internally after parsing a new message, and then use it for validating
against our masternode scoring.  (And in the future, we will use it
also to index our in-memory caches.)

domob1812 committed:
Pass CBlockIndex* to IsBlockPayeeValid.

IsBlockPayeeValid uses the block height and the previous block
CBlockIndex.  Instead of passing those two separately, just
pass the actual block's CBlockIndex*.

This is simpler, and it may also allow us to more directly compute
the scoring hash for the actual chain for validating masternode
payments, rather than doing it by height.

domob1812 committed:
Use seed hash as key for payee vote map.

Change the CMasternodePayments' internal mapMasternodeBlocks (which holds
the payment votes for each upcoming or recent block) to be keyed by
the block's scoring hash instead of height.  This makes it more robust
against reorgs.

domob1812 committed:
Compute scoring hash from CBlockIndex.

Instead of passing a block height and using chainActive, pass a
CBlockIndex and search in its explicit ancestor chain when computing
the scoring hash to be used for a block.  This further decouples the
masternode logic from chainActive as a global and makes it more robust
against reorgs.

The only remaining instance when we still use chainActive is for
computing the seed hash after parsing a CMasternodePaymentWinner
message from the wire.  In this case, we only have a block height,
so using chainActive as context is the best we can do (unless we
would want to change the wire protocol itself).

domob1812 committed:
Track payment votes more explicitly.

Instead of tracking votes for payment winners "indirectly" through
the height that each masternode has voted for last, just explicitly
keep track of the votes in the block payee data structure as a set
of masternodes who did vote.  (There will be only 10 anyway, so it
does not use up a lot of memory.)

This is simpler, more robust, and fixes potentially weird behaviour
when voting messages are received out-of-order (i.e. not in increasing
order of block heights).

domob1812 commented:
With this change applied, `divid` still successfully performs a masternode sync for me, accepts the winners / votes from the network, and in particular also never logs about a masternode "not being in the top" (which would mean a change with respect to how the scores are calculated).

galpHub commented:
Hey @domob1812 
This PR has a conflict after merging in the other masternode changes. Would appreciate if you could take a look :)

domob1812 commented:
> Hey @domob1812
> This PR has a conflict after merging in the other masternode changes. Would appreciate if you could take a look :)

Fixed.  I've also fixed a minor bug that broke the new `mnoperation.py` regtest (which now succeeds).  The issue was that `IsBlockPayeeValid` now uses the scoring hash internally, and that fails for blocks up to height 100 because it would be based on a block hash before genesis.  I've now simply made `IsBlockPayeeValid` skip masternode payment verification up to height 100 - on mainnet this does not matter anymore anyway, but it allows regtests to work fine.

galpHub commented on divi/src/masternode-payments.h:
```diff
@@ -112,6 +139,8 @@ class CMasternodePaymentWinner
         READWRITE(nBlockHeight);
         READWRITE(payee);
         READWRITE(vchSig);
+        if (ser_action.ForRead())
+            ComputeSeedHash();
```
Deserialization on read invoking the computation of the seed-hash could potentially result in a DDoS vector since the seed hash computation isn't cached. Better perhaps to just readwrite and soon after the start of the CMasternodePayments::ProcessMessageMasternodePayments method (for the case on mnwinners) you could have a public bool Verify() method that checks the hash perhaps as part of determining its validity.

Depending on the other changes this might not be releveant.

domob1812 replied:
This is a good point.  The only place where deserialisation happens is `ProcessMessageMasternodePayments`, but there are some more checks after the deserialisation in that function which we might as well do (and weed out some DoS attemps perhaps) before computing the hash.  I'll prepare an update that does the hash computation explicitly only right before it is needed (i.e. matching the behaviour before this PR).
